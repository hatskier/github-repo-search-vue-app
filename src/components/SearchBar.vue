<template>
  <div>
    <v-menu
      v-model="menu"
      bottom
      :offset-y="true"
      :disabled="selectableOptions.length === 0"
    >
      <template v-slot:activator="{ on, attrs }">
        <v-text-field
          v-bind="attrs"
          v-on="on"
          class="pa-5 pb-1 mb-4"
          outlined
          label="Search github repositories | click to select from saved"
          v-model="searchQuery"
          clearable
          hide-details
          :loading="loading"
          @keyup.enter.native="performSearch"
          append-icon="search"
          @click:append="performSearch"
        ></v-text-field>
      </template>
      <v-list>
        <v-list-item
          v-for="(term, index) in selectableOptions"
          :key="index"
          @click="selectTerm(term)"
        >
          <v-list-item-title>{{ term }}</v-list-item-title>

          <v-list-item-icon>
            <v-icon>chevron_right</v-icon>
          </v-list-item-icon>
        </v-list-item>
      </v-list>
    </v-menu>
    
  </div>
</template>

<script>
export default {
  name: 'SearchBar',
  data() {
    return {
      searchQuery: '',
      menu: false,
    }
  },
  props: {
    loading: Boolean,
    selectableOptions: Array,
  },
  methods: {
    selectTerm(selectedTerm) {
      this.searchQuery = selectedTerm
      this.performSearch()
    },
    performSearch() {
      this.$emit('search', this.searchQuery)
    },
    hideMenu() {
      this.menu = false
    }
  },
  watch: {
    searchQuery() {
      this.hideMenu()
    }
  },
}
</script>

<style scoped>

div.v-text-field__details {
  display: none;
}

</style>