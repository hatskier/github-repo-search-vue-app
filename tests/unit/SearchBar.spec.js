// Libraries
import Vuetify from 'vuetify'

// Components
import SearchBar from '@/components/SearchBar.vue'

// Utilities
import { mount, createLocalVue } from '@vue/test-utils'
import './setup'

const localVue = createLocalVue()

describe('GithubRepoCard.vue', () => {
  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  const mountFunction = options => {
    return mount(SearchBar, {
      localVue,
      vuetify,
      ...options,
    })
  }

  it('should indicate loading', () => {
    const wrapper = mountFunction({
      propsData: {
        loading: true,
        selectableOptions: [],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
    
    const inputDiv = wrapper.find('div.v-input.v-text-field')
    expect(inputDiv.classes()).toContain('v-input--is-loading')
  })

  it('should not indicate loading', () => {
    const wrapper = mountFunction({
      propsData: {
        loading: false,
        selectableOptions: [],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
    
    const inputDiv = wrapper.find('div.v-input.v-text-field')
    expect(inputDiv.classes()).not.toContain('v-input--is-loading')
  })

  it('should emit search event on search icon click', () => {
    const searchTerm = 'Test search term'

    const wrapper = mountFunction({
      propsData: {
        loading: false,
        selectableOptions: [],
      },
    })

    expect(wrapper.emitted().search).toBeFalsy()
    wrapper.find('[data-cy="search-input"]').setValue(searchTerm)
    wrapper.find('.v-input__icon--append > button').trigger('click')
    expect(wrapper.emitted().search).toHaveLength(1)
    expect(wrapper.emitted().search[0][0]).toBe(searchTerm)
  })

  it('should emit search event on enter keyup', () => {
    const searchTerm = 'Test search term'

    const wrapper = mountFunction({
      propsData: {
        loading: false,
        selectableOptions: [],
      },
    })

    expect(wrapper.emitted().search).toBeFalsy()
    wrapper.find('[data-cy="search-input"]').setValue(searchTerm)
    wrapper.find('[data-cy="search-input"]').trigger('keyup.enter')
    expect(wrapper.emitted().search).toHaveLength(1)
    expect(wrapper.emitted().search[0][0]).toBe(searchTerm)
  })
})
