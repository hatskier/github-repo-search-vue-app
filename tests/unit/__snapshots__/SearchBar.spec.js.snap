// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubRepoCard.vue should indicate loading 1`] = `
<div>
  <div class="v-menu">
    <div class="v-input pa-5 pb-1 mb-4 v-input--hide-details v-input--is-loading theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><label for="input-3" class="v-label theme--light" style="left: 0px; position: absolute;">Search github repositories | click to select from saved</label><input data-cy="search-input" role="button" aria-haspopup="true" aria-expanded="false" id="input-3" type="text"></div>
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--clear"><button disabled="disabled" type="button" aria-label="clear icon" class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--light"></button></div>
          </div>
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append"><button type="button" aria-label="append icon" class="v-icon notranslate v-icon--link material-icons theme--light">search</button></div>
          </div>
          <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="v-progress-linear v-progress-linear--absolute theme--light" style="height: 2px;">
            <div class="v-progress-linear__background primary" style="opacity: 0.3; left: 0%; width: 100%;"></div>
            <div class="v-progress-linear__buffer"></div>
            <div class="v-progress-linear__indeterminate v-progress-linear__indeterminate--active">
              <div class="v-progress-linear__indeterminate long primary"></div>
              <div class="v-progress-linear__indeterminate short primary"></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <!---->
  </div>
</div>
`;

exports[`GithubRepoCard.vue should not indicate loading 1`] = `
<div>
  <div class="v-menu">
    <div class="v-input pa-5 pb-1 mb-4 v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><label for="input-11" class="v-label theme--light" style="left: 0px; position: absolute;">Search github repositories | click to select from saved</label><input data-cy="search-input" role="button" aria-haspopup="true" aria-expanded="false" id="input-11" type="text"></div>
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--clear"><button disabled="disabled" type="button" aria-label="clear icon" class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--light"></button></div>
          </div>
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append"><button type="button" aria-label="append icon" class="v-icon notranslate v-icon--link material-icons theme--light">search</button></div>
          </div>
        </div>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
